kmath-kotlingrad/src/test/kotlin/kscience/kmath/kotlingrad/AdaptingTests.kt

```diff
@@ -0,0 +1,66 @@
+package kscience.kmath.kotlingrad
+
+import edu.umontreal.kotlingrad.experimental.*
+import kscience.kmath.asm.compile
+import kscience.kmath.ast.MstAlgebra
+import kscience.kmath.ast.MstExpression
+import kscience.kmath.ast.parseMath
+import kscience.kmath.expressions.invoke
+import kscience.kmath.operations.RealField
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertTrue
+import kotlin.test.fail
+
+internal class AdaptingTests {
+    private val proto: DReal = DoublePrecision.prototype
+
+    @Test
+    fun symbol() {
+        val c1 = MstAlgebra.symbol("x")
+        assertTrue(c1.toSVar(proto).name == "x")
+        val c2 = "kitten".parseMath().toSFun(proto)
+        if (c2 is SVar) assertTrue(c2.name == "kitten") else fail()
+    }
+
+    @Test
+    fun number() {
+        val c1 = MstAlgebra.number(12354324)
+        assertTrue(c1.toSConst<DReal>().doubleValue == 12354324.0)
+        val c2 = "0.234".parseMath().toSFun(proto)
+        if (c2 is SConst) assertTrue(c2.doubleValue == 0.234) else fail()
+        val c3 = "1e-3".parseMath().toSFun(proto)
+        if (c3 is SConst) assertEquals(0.001, c3.value) else fail()
+    }
+
+    @Test
+    fun simpleFunctionShape() {
+        val linear = "2*x+16".parseMath().toSFun(proto)
+        if (linear !is Sum) fail()
```

Comment:
Better to use AssertFalse

Reply:
No. `fail` is better since it returns Nothing, so Kotlin DFA makes smart-cast of `linear` to `Sum`